vhost-user-backend/src/bitmap.rs

```diff
@@ -0,0 +1,603 @@
+unsafe impl Send for MmapLogReg {}
+
+// SAFETY: Sync is not automatically implemented because the raw pointer.
+// `MmapLogReg` doesn't have any interior mutability and all access to `&AtomicU8`
+// are done through atomic operations.
+unsafe impl Sync for MmapLogReg {}
+
+impl MmapLogReg {
+    #![allow(dead_code)]
+    // Note: We could try to adjust the mapping area to only cover the memory region, but
+    // the region's starting address is not guarantee to be LOG_WORD_SIZE-page aligned
+    // which makes the implementation needlessly cumbersome.
+    // Note: The specification does not define whether the offset must be page-aligned or not.
+    // But, since we are receiving the offset from the frontend to be used to call mmap,
+    // we assume it is properly aligned (currently, qemu always send a 0 offset).
+    pub(crate) fn from_file(fd: BorrowedFd, offset: u64, len: u64) -> io::Result<Self> {
+        let offset: isize = offset.io_try_into()?;
+        let len: usize = len.io_try_into()?;
+
+        // Let's uphold the safety contract for `std::ptr::offset()`.
+        if len > isize::MAX as usize {
+            return Err(io::Error::from(io::ErrorKind::InvalidData));
+        }
+
+        // SAFETY: `fd` is a valid file descriptor and we are not using `libc::MAP_FIXED`.
+        let addr = unsafe {
+            libc::mmap(
+                ptr::null_mut(),
+                len as libc::size_t,
+                libc::PROT_READ | libc::PROT_WRITE,
+                libc::MAP_SHARED,
+                fd.as_raw_fd(),
+                offset as libc::off_t,
+            )
+        };
+
+        if addr == libc::MAP_FAILED {
+            return Err(io::Error::last_os_error());
+        }
+
+        Ok(Self {
+            addr: addr as *const AtomicU8,
+            len,
+        })
+    }
+
+    fn len(&self) -> usize {
+        self.len
+    }
+}
+
+impl Index<usize> for MmapLogReg {
+    type Output = AtomicU8;
+
+    // It's ok to get a reference to an atomic value.
+    fn index(&self, index: usize) -> &Self::Output {
+        assert!(index < self.len);
+        // Note: Instead of `&*` we can use `AtomicU8::from_ptr()` as soon it gets stabilized.
+        // SAFETY: `self.addr` + `index` doesn't wrap around and is contained within the mapped
+        // memory region.
+        unsafe { &*self.addr.add(index) }
+    }
+}
+
+impl Drop for MmapLogReg {
+    fn drop(&mut self) {
+        // SAFETY: `addr` is properly aligned, also we are sure that this is the
+        // last reference alive and/or we have an exclusive access to this object.
+        unsafe {
+            libc::munmap(self.addr as *mut libc::c_void, self.len as libc::size_t);
+        }
+    }
+}
+
+trait IoTryInto<T: TryFrom<Self>>: Sized {
+    fn io_try_into(self) -> io::Result<T> {
+        self.try_into()
+            .map_err(|_| io::Error::from(io::ErrorKind::InvalidData))
+    }
+}
+
+impl<TySrc, TyDst> IoTryInto<TyDst> for TySrc where TyDst: TryFrom<TySrc> {}
```

Comment:
`io::Error` allows storing (more or less) arbitrary error objects within, so we don’t need to discard the old error.  This could be:
```rust
trait IoTryInto<T: TryFrom<Self>>: Sized {
    fn io_try_into(self) -> io::Result<T>;
}

impl<TySrc, TyDst: TryFrom<TySrc>> IoTryInto<TyDst> for TySrc 
where
    <TyDst as TryFrom<TySrc>>::Error: Send + Sync + std::error::Error + 'static,
{                                                                                                                                                                               
    fn io_try_into(self) -> io::Result<TyDst> {
        self.try_into()
            .map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e))
    }                                 
}
```
But maybe the whole error is discarded anyway and it only matters that there was an error, then this may be overkill.